lightly/models/simsiam.py

```diff
@@ -0,0 +1,159 @@
+import torch
+import torch.nn as nn
+
+from lightly.models.resnet import ResNetGenerator
+from lightly.models.batchnorm import get_norm_layer
+from lightly.models._loader import _StateDictLoaderMixin
+
+
+def _prediction_mlp(in_dims: int = 2048, 
+                    h_dims: int = 512, 
+                    out_dims: int = 2048) -> nn.Sequential:
+    """Prediction MLP. The original paper's implementation has 2 layers, with 
+    BN applied to its hidden fc layers but no BN or ReLU on the output fc layer.
+
+    Note that the hidden dimensions should be smaller than the input/output 
+    dimensions (bottleneck structure). The default implementation using a 
+    ResNet50 backbone has an input dimension of 2048, hidden dimension of 512, 
+    and output dimension of 2048
+
+    Args:
+        in_dims:
+            Input dimension of the first linear layer.
+        h_dims: 
+            Hidden dimension of all the fully connected layers (should be a
+            bottleneck!)
+        out_dims: 
+            Output Dimension of the final linear layer.
+
+    Returns:
+        nn.Sequential:
+            The projection head.
+    """
+    l1 = nn.Sequential(nn.Linear(in_dims, h_dims),
+                       nn.BatchNorm1d(h_dims),
+                       nn.ReLU(inplace=True))
+
+    l2 = nn.Linear(h_dims, out_dims)
+
+    prediction = nn.Sequential(l1, l2)
+    return prediction
+
+
+def _projection_mlp(in_dims: int = 2048, 
+                    h_dims: int = 2048, 
+                    out_dims: int = 2048) -> nn.Sequential:
+    """Projection MLP. The original paper's implementation has 3 layers, with 
+    BN applied to its hidden fc layers but no ReLU on the output fc layer.
+
+    Args:
+        in_dims:
+            Input dimension of the first linear layer.
+        h_dims: 
+            Hidden dimension of all the fully connected layers.
+        out_dims: 
+            Output Dimension of the final linear layer.
+
+    Returns:
+        nn.Sequential:
+            The projection head.
+    """
+    l1 = nn.Sequential(nn.Linear(in_dims, h_dims),
+                       nn.BatchNorm1d(h_dims),
+                       nn.ReLU(inplace=True))
+
+    l2 = nn.Sequential(nn.Linear(h_dims, h_dims),
+                       nn.BatchNorm1d(h_dims),
+                       nn.ReLU(inplace=True))
+
+    l3 = nn.Sequential(nn.Linear(h_dims, out_dims),
+                       nn.BatchNorm1d(out_dims))
```

Comment:
I'm not sure whether we should add here or on line 33 (before linear layer) another ReLU activation. Currently, we have

- linear
- batch_norm
- linear

two linear layers without non-linearity between seem odd to me. 

I would add ReLU at line 33. That way we don't have a sparse output after the projection head.

Reply:
That's true it is not typical. However this would match the implementation described in [the paper (pg 3)](https://arxiv.org/pdf/2011.10566.pdf): 

> _Projection MLP_. The projection MLP _(f)_ has BN applied to each fully-connected (fc) layer, including its output fc. Its **output fc has no ReLU**. The hidden fc is 2048 d. This MLP has 3 layers. 
>
> _Prediction MLP._ The prediction MLP _(h)_ has BN applied to its hidden fc layers.   Its output fc does not have BN (ablation in Sec. 4.4) or ReLU. This MLP has 2 layers. The dimension of _h_’s input and output (_z_ and _p_) is d=2048, and _h_’s hidden layer’s dimension is 512, making _h_ a bottleneck structure (ablation in supplement).

This could be worth experimenting with to see the difference in performance?

Reply:
Got this reply from Kaiming:
Dear Igor,

 

Thank you for your interest!

 

The predictor head by default is: Linear->BN->ReLU->Linear. Also note the dimension is 2048->512->2048. And you are right, there is no ReLU between the output of projection head and the input of prediction head. This is because this embedding (output of projection head) is expected to be normalized to the unit sphere, so ReLU is not desired.

 

Regards,

Kaiming